### PR TITLE
Curl: Fix #curl_args options

### PR DESCRIPTION
**Pull request by samford**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

`Utils::Curl#curl_args` seems like it intends to accept `:connect_timeout`, `:max_time`, and `:retry_max_time` options but the code that adds arguments for these values uses variables that aren't defined in this scope (`connect_timeout`, `max_time`, and `retry_max_time` respectively). The related code gives an error like ```NameError (undefined local variable or method `connect_timeout' for main:Object)``` when these options are provided to `#curl_args`.

~This PR updates this code to use the `options` value for each of these, so this will work as intended. This also modifies the inline conditions to ensure these values are numeric, as a non-numeric value would cause `#round` (and `curl`) to error.~

After discussion, this PR updates `#curl_args` to use named parameters instead of `**options`, which resolves the undefined variable issue. This also updates related code in Homebrew/brew to use these parameters instead of passing related args like `--connect-timeout`.

---

Just as an aside, I've extracted this commit from a local branch that I'll be creating a PR for soon. At the moment, I'm working on creating PRs for necessary changes that aren't strictly topical, as a way of simplifying the forthcoming PR.

### CODE REVIEW

**Review by Rylan12 — APPROVED** [review]
Makes sense, good catch! Thanks!

File: Library/Homebrew/utils/curl.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/curl.rb:61]: Seems a bit strange that this will simply silently discard the value of e.g. `options[:connect_timeout] = "3"`. Feels more Ruby-like to either convert it to the type we expect or `raise` if it's the wrong type. Also, while I'm being pedantic, `**options` feels A Bit Ugly now we have kwargs and Sorbet so would be better refactored to use them.

Comment (samford) [reply Library/Homebrew/utils/curl.rb:61]: > Seems a bit strange that this will simply silently discard the value of e.g. `options[:connect_timeout] = "3"`. Feels more Ruby-like to either convert it to the type we expect or `raise` if it's the wrong type.

I was getting that feeling, too. I'll add some logic that `raise`s on a non-numeric value in the next push, as that's an easy solution.

---

> Also, while I'm being pedantic, `**options` feels A Bit Ugly now we have kwargs and Sorbet so would be better refactored to use them.

Just to make sure I understand, are you suggesting making these options into named parameters, like so:

```ruby
sig {
  params(
    extra_args:      T::Array[T.untyped],
    connect_timeout: T.any(Integer, Float, NilClass),
    max_time:        T.any(Integer, Float, NilClass),
    retry_max_time:  T.any(Integer, Float, NilClass),
    options:         T.nilable(T::Hash[Symbol, T.untyped]),
  ).returns(T::Array[T.untyped])
}
def curl_args(*extra_args, connect_timeout: nil, max_time: nil, retry_max_time: nil, **options)
```

`retry` is a Ruby keyword, so we still have to use `options[:retry]` to allow us to use `retry: false` when calling `#curl_args`.

Also, we haven't enabled typechecking for `Utils::Curl` yet (there are ~40 issues to resolve first), so the signature won't do anything right now. Enforcing the correct type for these arguments within `#curl_args` (as mentioned above) is necessary anyway, as args/options will sometimes come from places where `brew typecheck` doesn't look (e.g., a `livecheck` block in a formula/cask).

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/curl.rb:61]: > Just to make sure I understand, are you suggesting making these options into named parameters, like so:

Yes!

> `retry` is a Ruby keyword, so we still have to use `options[:retry]` to allow us to use `retry: false` when calling `#curl_args`.

👍🏻 but I think it may be worth a deprecation (or rename if this a private API which I think it is?) for this reason.

Comment (samford) [reply Library/Homebrew/utils/curl.rb:61]: The `retry` option is only used in `Livecheck::Strategy` at the moment (I added it specifically for livecheck), so it would be easy to rename it to something that can be used as a named parameter. `#curl_args` is listed as private in the docs, so renaming sounds fine.

Do you have any ideas for an alternative name for `retry`? When we were discussing this in the related PR, alternative names felt clunky (e.g., `retry_enabled`, `should_retry`, etc.) and `retry` was technically possible with the `options` setup, so we went that route. If not, I'll think on this later and see if I can come up with anything reasonable.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/curl.rb:61]: > The `retry` option is only used in `Livecheck::Strategy` at the moment (I added it specifically for livecheck), so it would be easy to rename it to something that can be used as a named parameter. `#curl_args` is listed as private in the docs, so renaming sounds fine.

👍🏻 

> should_retry

This sounds good given keyword.



> Thinking about this some more, one other way to approach this would be to replace `retry` with a `retries` parameter (defaulting to `Homebrew::EnvConfig.curl_retries`) and we would only add the `--retry` arg when `retries.positive?` is true. This would allow us to use `retries: 0` to avoid the default `--retry`.

This also works. I like the elegance here.

**Comment by samford** [comment]
The latest push reworks `#curl_args` to use named parameters and replaces the `retry` parameter with `retries`. This also adds logic to `raise` a `TypeError` when related arguments aren't an appropriate type.

I've gone through the codebase to replace use of `--connect-timeout`, `--max-time`, and `--retry-max-time` with the named parameters. I also refactored some areas of `Utils::Curl` to make parameter/argument order more consistent.

File: Library/Homebrew/utils/curl.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/utils/curl.rb:87]: I think given Sorbet and this being a private API it might be OK to just omit these type errors? Thoughts?

**Review by MikeMcQuaid — APPROVED** [review]
Great work @samford! One thought but fine to merge as-is, your call.

File: Library/Homebrew/utils/curl.rb
Comment (samford) [reply Library/Homebrew/utils/curl.rb:87]: By default, `sorbet-runtime` is only enabled in `brew tests` at the moment (via [`ENV["HOMEBREW_SORBET_RUNTIME"] = "1"`](https://github.com/Homebrew/brew/blob/b253b304d925bef2582c60eda22342e35a8205ba/Library/Homebrew/dev-cmd/tests.rb#L94)), so Sorbet only handles the type errors in that context. Outside of `brew tests`, we would rely on these `raise` statements and that's why I left them in.

After this is merged, I'll be creating a PR to allow for configuration of `curl` args/options in `livecheck` blocks. This is necessary for a variety of use cases (configuring the user agent, `POST` requests, etc.) and the values will be passed through livecheck to `#curl_args`.

With this in mind, I think it's better to raise an error, so the contributor will know that something isn't working as expected when they run `brew livecheck` on the formula/cask. Otherwise, `#curl_args` would quietly omit the argument on a non-numeric value and that can be a problem if it doesn't cause the check to noticeably fail. Even if the check does fail as a result, it wouldn't be immediately clear why it isn't working as expected and these errors are useful in that regard as well.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/curl.rb:87]: > By default, `sorbet-runtime` is only enabled in `brew tests` at the moment (via [`ENV["HOMEBREW_SORBET_RUNTIME"] = "1"`](https://github.com/Homebrew/brew/blob/b253b304d925bef2582c60eda22342e35a8205ba/Library/Homebrew/dev-cmd/tests.rb#L94)), so Sorbet only handles the type errors in that context. Outside of `brew tests`, we would rely on these `raise` statements and that's why I left them in (after manually testing the behavior).

Sorbet will only handle runtime but it's still doing full-codebase static analysis. I think that's sufficient enough to remove the `TypeError`s (or can these values be set by non-maintainers?)

Comment (samford) [reply Library/Homebrew/utils/curl.rb:87]: My concern is about values provided indirectly to `#curl_args` from outside of the Homebrew/brew codebase (e.g., formulae/casks in first- and third-party taps). In the near future, we'll have `curl` args/options coming from a method in the `livecheck` block DSL but this can be addressed by doing the type checking in that method instead. We already do this with the other `livecheck` block DSL methods, so it makes sense to continue with that approach.

This would allow us to rely on Sorbet to do type checking for `#curl_args` but `Utils::Curl` is currently `typed: false` and there are ~40 issues to resolve before we can make it `true`. Until then, `brew typecheck` won't be enforcing argument types for `#curl_args` (and `sorbet-runtime` is only enabled for tests, so that won't help). [I can eventually work on making `Curl` typed as part of the other `Curl` work I have on the back burner but it won't be until I've finished with some forthcoming livecheck stuff (i.e., not soon).]

Removing the `raise` statements wouldn't be any worse than the status quo before this PR. If we do, we should also loosen the related conditions from `connect_timeout.is_a?(Numeric)` to `connect_timeout.present?`, so a bad value will cause a noticeable failure (in `#round` or from `curl`) instead of quietly omitting an argument.

This seems reasonable to me, so I'll update this accordingly.

Comment (MikeMcQuaid) [reply Library/Homebrew/utils/curl.rb]: > My concern is about values provided indirectly to `#curl_args` from outside of the Homebrew/brew codebase (e.g., formulae/casks in first- and third-party taps).

I think it's worth having these checks in the public methods rather than this private one so the backtrace makes more sense.

> This would allow us to rely on Sorbet to do type checking for `#curl_args` but `Utils::Curl` is currently `typed: false` and there are ~40 issues to resolve before we can make it `true`. Until then, `brew typecheck` won't be enforcing argument types for `#curl_args` (and `sorbet-runtime` is only enabled for tests, so that won't help). [I can eventually work on making `Curl` typed as part of the other `Curl` work I have on the back burner but it won't be until I've finished with some forthcoming livecheck stuff (i.e., not soon).]

👍🏻 


> Removing the `raise` statements wouldn't be any worse than the status quo before this PR. If we do, we should also loosen the related conditions from `connect_timeout.is_a?(Numeric)` to `connect_timeout.present?`, so a bad value will cause a noticeable failure (in `#round` or from `curl`) instead of quietly omitting an argument.

I defer to you here @samford, merge whatever you feel is best (including this PR as-is, if desired).

Comment (samford) [reply Library/Homebrew/utils/curl.rb]: > I think it's worth having these checks in the public methods rather than this private one so the backtrace makes more sense.

Agreed. The errors from the `livecheck` block method are likely to be more understandable to contributors/maintainers than something coming from `#curl_args` 👍

> I defer to you here @samford, merge whatever you feel is best (including this PR as-is, if desired).

I removed the `raise` statements and loosened the conditions from `#is_a?` to `#present?` in the latest push. I imagine this should be fine, as some of these options didn't even function before and no one noticed until now 😅

Switching to named parameters and adding a method signature will automatically produce better documentation for `#curl_args`, so folks will be able to see the expected type for these parameters. Even with `Utils::Curl` being `typed: false`, this is an improvement over the existing situation and it will improve further when we can switch to `typed: true` in the future.

**Comment by samford** [comment]
I think this is in good shape now, so I'm going to go ahead and merge. Thanks for the feedback, everyone!

**Comment by MikeMcQuaid** [comment]
Great work, thanks again @samford!

**Pull request merged**
